sys/net/gnrc/link_layer/csma_sender/gnrc_csma_sender.c

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright (C) 2015 INRIA
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ * @ingroup     net_csma_sender
+ * @file
+ * @brief       Implementation of the CSMA/CA helper
+ *
+ * @author      Kévin Roussel <Kevin.Roussel@inria.fr>
+ * @}
+ */
+
+#include <errno.h>
+#include <stdbool.h>
+
+#include "kernel.h"
+#include "xtimer.h"
+#include "net/gnrc/csma_sender.h"
+#include "net/gnrc.h"
+#include "net/netopt.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+/* For PRIu16 etc. */
+#include <inttypes.h>
+#endif
+
+
+/** @brief Current value for mac_min_be parameter */
+static uint8_t mac_min_be = MAC_MIN_BE_DEFAULT;
+
+/** @brief Current value for mac_max_be parameter */
+static uint8_t mac_max_be = MAC_MAX_BE_DEFAULT;
+
+/** @brief Current value for mac_max_csma_backoffs parameter */
+static uint8_t mac_max_csma_backoffs = MAC_MAX_CSMA_BACKOFFS_DEFAULT;
+
+
+/*--------------------- "INTERNAL" UTILITY FUNCTIONS ---------------------*/
+
+/**
+ * @brief choose an adequate random backoff period in microseconds,
+ *        from the given Backoff Exponent
+ *
+ * @param[in] be        Backoff Exponent for the computation of period
+ *
+ * @return              An adequate random backoff exponent in microseconds
+ */
+static inline uint32_t choose_backoff_period(int be)
+{
+    if (be < mac_min_be) {
+        be = mac_min_be;
+    }
+    if (be > mac_max_be) {
+        be = mac_max_be;
+    }
+    uint32_t max_backoff = ((1 << be) - 1) * A_UNIT_BACKOFF_PERIOD_MICROSEC;
+
+    uint32_t period = genrand_uint32() % max_backoff;
+    if (period < A_UNIT_BACKOFF_PERIOD_MICROSEC) {
+        period = A_UNIT_BACKOFF_PERIOD_MICROSEC;
+    }
+
+    return period;
+}
+
+
+/**
+ * @brief Perform a CCA and send the given packet if medium is available
+ *
+ * @param[in] device    netdev device, needs to be already initialized
+ * @param[in] data      pointer to the data in the packet buffer
```

Comment:
Yes indeed: it must be a complete 802.15.4-compliant frame, ready to be sent to the radio transceiver...
I'll add this precision in the documentation.